transport/httptransport/http_downloader.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	// add request headers
+	for name, val := range d.transfer.tInfo.Headers {
+		req.Header.Set(name, val)
+	}
+
+	// calculate range boundaries taking into account bytes that have been already downloaded
+	fi, err := os.Stat(d.chunkFile)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return &httpError{error: fmt.Errorf("failed to read file size: %w", err)}
+	}
+	var offset int64
+	if fi != nil {
+		offset = fi.Size()
+	}
+
+	rangeStart := d.rangeStart + offset
+
+	// the file might have been already fully downloaded, nothing to do here
+	if rangeStart >= d.rangeEnd {
+		return nil
+	}
+
+	toRead := d.rangeEnd - rangeStart
+
+	// add range req to start reading from the last byte we have in the output file
+	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", rangeStart, d.rangeEnd))
+	// init the request with the transfer context
+	req = req.WithContext(d.ctx)
+
+	duid := d.transfer.dealInfo.DealUuid
+	d.transfer.dl.Infow(duid, "sending http request", "toRead", toRead, "range-rq", req.Header.Get("Range"))
+
+	// send http request and validate response
+	resp, err := d.transfer.client.Do(req)
+	if err != nil {
+		return &httpError{error: fmt.Errorf("failed to send  http req: %w", err)}
+	}
+	// we should either get back a 200 or a 206 -> anything else means something has gone wrong and we return an error.
+	defer resp.Body.Close() // nolint
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+		return &httpError{
+			error: fmt.Errorf("http req failed: code: %d, status: %s", resp.StatusCode, resp.Status),
+			code:  resp.StatusCode,
+		}
+	}
+
+	// create chunk file if it doesn't exist
+	of, err := os.OpenFile(d.chunkFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return &httpError{error: fmt.Errorf("failed to open chunk file for write: %w", err)}
+	}
+	defer of.Close()
+
+	//  start reading the response stream `readBufferSize` at a time using a limit reader so we only read as many bytes as we need to.
+	buf := make([]byte, readBufferSize)
+	limitR := io.LimitReader(resp.Body, toRead)
+	var chunkBytesReceived int64
+	for {
+		if d.ctx.Err() != nil {
+			d.transfer.dl.LogError(duid, "stopped reading http response: context canceled", d.ctx.Err())
+			return &httpError{error: d.ctx.Err()}
+		}
+		nr, readErr := limitR.Read(buf)
+
+		// if we read more than zero bytes, write whatever read.
+		if nr > 0 {
+			nw, writeErr := of.Write(buf[0:nr])
+
+			// if the number of read and written bytes don't match -> something has gone wrong, abort the http req.
+			if nw < 0 || nr != nw {
+				if writeErr != nil {
+					return &httpError{error: fmt.Errorf("failed to the chunk file: %w", writeErr)}
+				}
+				return &httpError{error: fmt.Errorf("read-write mismatch writing to the chunk file, read=%d, written=%d", nr, nw)}
+			}
+			chunkBytesReceived += int64(nw)
+			totalBytesReceived := atomic.AddInt64(&d.transfer.nBytesReceived, int64(nw))
```

Comment:
We should check that wherever transfer.nBytesReceived is being accessed, it's accessed with atomic.LoadInt64().
Maybe it would make sense to add a method to transfer (eg `addBytesReceived(delta int64)`) that
- updates the nBytesReceived (with an exclusive lock)
- emits an event with the new total bytes received

Reply:
Good point - added